### PR TITLE
auto-restart on file changes to services

### PR DESCRIPTION
**Pull request by hugobho**
services auto-restart on file additions, changes or deletions
feature tests written under "run.feature"
#1

### CODE REVIEW

File: features/auto_restart.feature [hunk 1]
```diff
@@ -0,0 +1,20 @@
+Feature: auto-restart services on file changes
+
+  As a developer editing an Exosphere application
+  I want that any changes I make to the source code gets loaded immediately into the runtime
```
Comment (kevgo) [inline features/auto_restart.feature]: get~~s~~

File: features/auto_restart.feature [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+  So that I don't waste time manually restarting services all the time, or setting up service-specific auto-reloading.
+
+  Rules:
+  - anytime a file in a service directory is created, updated, or deleted, "exo run" restarts the respective service
+  - when a service is restarted, the logger outputs: "exo run: restarting service '<service name>'"
+  
+  Scenario: a file watcher reboots after changes to source code are made
+    Given a set-up "test" application
+    When starting "exo run" in this application's directory
+    And waiting until I see "application ready" in the terminal
```
Comment (kevgo) [inline features/auto_restart.feature]: This spec describes how restarting works in real life. 
The minutiae of how Exosphere applications are started are not relevant for this.
The first 3 lines can therefore be summarized to `Given a running "test" application".

File: features/auto_restart.feature [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+    When I add a file "new-file.js" to the "web-server" service
```
Comment (kevgo) [inline features/auto_restart.feature]: Thinking in terms of explaining how restarting services works, it doesn't matter who adds the file (I or some program). What matters is that the file is created.
So let's say that: `When a file "web-server/new-file.js" gets created`

File: features/steps/steps.ls [hunk 4]
```diff
@@ -128,6 +128,22 @@ module.exports = ->
     request 'http://localhost:4000', done
 
 
+  @When /^I add a file "([^"]*)" to the "([^"]*)" service$/ (new-filename, service, done) ->
+    service-dir = path.join process.cwd!, 'tmp', @app-name, service
+    new-file-path = path.join service-dir, new-filename
+    fs.open new-file-path, 'w', done
```
Comment (kevgo) [inline features/steps/steps.ls]: If you use the sync method here, you don't need the `done` callback.

```ls
fs.open-sync new-file-path, 'w'

File: features/steps/steps.ls [hunk 5]
```diff
@@ -128,6 +128,22 @@ module.exports = ->
...
+
+  @When /^I change the file "([^"]*)" in the "([^"]*)" service$/ (filename, service, done) ->
+    service-dir = path.join process.cwd!, 'tmp', @app-name, service
+    file-path = path.join service-dir, filename
+    file-data = "var hello = \"hello\""
+    fs.write-file file-path, file-data, done
```
Comment (kevgo) [inline features/steps/steps.ls]: Maybe use https://nodejs.org/api/fs.html#fs_fs_appendfilesync_file_data_options to make more clear that the existing file is just changed?

File: features/steps/steps.ls [hunk 6]
```diff
@@ -128,6 +128,22 @@ module.exports = ->
...
+
+  @When /^I delete the file "([^"]*)" in the "([^"]*)" service$/ (filename, service, done) ->
+    service-dir = path.join process.cwd!, 'tmp', @app-name, service
```
Comment (kevgo) [inline features/steps/steps.ls]: I think you can inline the variable `service-dir` in all 3 steps above

File: features/steps/steps.ls [hunk 7]
```diff
@@ -214,3 +230,26 @@ module.exports = ->
 
   @Then /^the "([^"]*)" service replies with a "([^"]*)" message$/, (arg1, arg2, done) ->
     done!
+
+  @Then /^the "([^"]*)" service is restarted( again)?$/, (service, again, done) ->
+
```
Comment (kevgo) [inline features/steps/steps.ls]: Remove empty line

File: features/steps/steps.ls [hunk 8]
```diff
@@ -214,3 +230,26 @@ module.exports = ->
...
+    unless again 
+      #build expected output if service was restarted for the first time
+      @restarting-text = "exo run: restarting service '#{service}'\n"
+      output-lines = @process.full-output!.split /\r?\n/
+      for i from 0 til output-lines.length
```
Comment (kevgo) [inline features/steps/steps.ls]:
```ls
for line in @process.full-output!.split /\r?\n/
  if line.includes

File: package.json [hunk 9]
```diff
@@ -6,14 +6,15 @@
     "exo": "bin/exo"
   },
   "dependencies": {
-    "async": "1.5.2",
+    "async": "^1.5.2",
```
Comment (kevgo) [inline package.json]: Remove the `^` everywhere here.

File: src/commands/run/app-runner.ls [hunk 10]
```diff
@@ -16,6 +17,10 @@ require! {
 class AppRunner extends EventEmitter
 
   (@app-config) ->
+    @watcher = chokidar.watch _ , {
```
Comment (kevgo) [inline src/commands/run/app-runner.ls]: What does the underscore mean here? Shouldn't this be "." or something like that?

File: src/commands/run/app-runner.ls [hunk 11]
```diff
@@ -16,6 +17,10 @@ require! {
...
+      ignoreInitial: true,
+      ignored: /.*\/node_modules\/.*/
```
Comment (kevgo) [inline src/commands/run/app-runner.ls]:
```ls
@watcher = chokidar.watch '.', 
                          ignore-initial: yes,
                          ignored: /.*\/node_modules\/.*/

File: src/commands/run/app-runner.ls (hunk 10 above)
Comment (kevgo) [inline src/commands/run/app-runner.ls]: Shouldn't each ServiceRunner watch their own directory? An app can use services outside of its directory.

File: features/auto_restart.feature [hunk 12]
```diff
@@ -0,0 +1,20 @@
...
+    Then the "web" service is restarted
+    When I change the file "new-file.js" in the "web-server" service
+    Then the "web" service is restarted again
+    When I delete the file "new-file.js" in the "web-server" service
+    Then the "web" service is restarted again
```
Comment (kevgo) [inline features/auto_restart.feature]: There are a number of scenarios missing here:

```cucumber
Scenario: file change in an internal service

Scenario: file change in an external service

File: src/commands/run/service-runner.ls [hunk 13]
```diff
@@ -11,25 +12,46 @@ require! {
 
 class ServiceRunner extends EventEmitter
 
-  (@name, @config) ->
+  (@name, @config, @watcher) ->
     @service-config = yaml.safe-load fs.readFileSync(path.join(@config.root, 'service.yml'), 'utf8')
     @service-config.messages.sends ?= []
     @service-config.messages.receives ?= []
+    @watcher @config.root
+      .on('add', (watch-path) ~>
```
Comment (kevgo) [inline src/commands/run/service-runner.ls]: Omit parentheses:

```ls
@watcher @config.root
  .on 'add', (watch-path) ~>
    #...
  .on 'change', (...

File: src/commands/run/service-runner.ls [hunk 14]
```diff
@@ -11,25 +12,46 @@ require! {
...
     port-reservation.get-port N (port) ~>
       @config.SERVICE_NAME = @name
       @config.EXORELAY_PORT = port
-      new ObservableProcess(@_create-command(@service-config.startup.command)
+      @server-process = new ObservableProcess(@_create-command(@service-config.startup.command)
                             cwd: @config.root,
                             env: @config
                             console: log: @_log, error: @_log)
-        ..on 'ended', ~> throw new Error "Service '#{@name}' crashed"
+        ..on 'ended', (exitCode, killed) ~> throw new Error "Service '#{@name}' crashed" unless killed is true
```
Comment (kevgo) [inline src/commands/run/service-runner.ls]: `exit-code`

Comment (kevgo) [inline src/commands/run/service-runner.ls]: `unless killed`

File: src/commands/run/service-runner.ls [hunk 15]
```diff
@@ -11,25 +12,46 @@ require! {
...
         ..wait @service-config.startup['online-text'], ~>
           @emit 'online', @name
           done!
 
+  _restart: ~>
+    @server-process.kill()
```
Comment (kevgo) [inline src/commands/run/service-runner.ls]: @server-process.kill!

Comment (kevgo) [inline src/commands/run/service-runner.ls]: I suggest waiting until the old process has ended before creating a new one here.

**Comment by hugobho** [comment]
@kevgo

File: example-apps/external-service/test/application.yml [hunk 16]
```diff
@@ -0,0 +1,11 @@
+name: Example application
+description: Demonstrates basic Exosphere application startup
+version: '1.0'
+
+services:
+  web:
+    location: ../web-server
+  users:
+    location: ./mongo-service
+  dashboard:
```
Comment (kevgo) [inline example-apps/external-service/test/application.yml]: Do you really need 3 services here? Can you delete the dashboard service to make this test faster?

File: features/auto_restart.feature [hunk 17]
```diff
@@ -0,0 +1,27 @@
...
+  So that I don't waste time manually restarting services all the time, or setting up service-specific auto-reloading.
+
+  Rules:
+  - anytime a file in a service directory is created, updated, or deleted, "exo run" restarts the respective service
+  - when a service is restarted, the logger outputs: "exo run: restarting service '<service name>'"
+
+  Scenario: file change in an internal service
+    Given a running "test" application
+    When a file "web-server/new-file.js" gets created
+    Then the "web" service is restarts
```
Comment (kevgo) [inline features/auto_restart.feature]: This is weird grammar. Better: 
> Then the "web" service gets restarted

File: features/auto_restart.feature [hunk 18]
```diff
@@ -0,0 +1,27 @@
...
+    When the file "web-server/new-file.js" gets changed
+    Then the "web" service is restarts again
+    When the file "web-server/new-file.js" gets deleted
+    Then the "web" service is restarts again
+
+  Scenario: file change in an external service
+    Given a running "external-service/test" application
```
Comment (kevgo) [inline features/auto_restart.feature]: Rather than creating a completely new app with 3 services, can you just create a new "external-service" example app with one service in it, which is external?

Comment (kevgo) [inline features/auto_restart.feature]: Also, please try to keep the services as simple as possible in the example app. No need for real data handling or rendering HTML, just let them print some data.

File: features/steps/steps.ls [hunk 19]
```diff
@@ -307,6 +340,12 @@ module.exports = ->
     done!
 
 
+  @Then /^the "([^"]*)" service is restarts( again)?$/, (service, again, done) ->
+    @process.reset-output-streams!
+    expected-text = "exorun  '#{service}' is running using exorelay port"
```
Comment (kevgo) [inline features/steps/steps.ls]: Inline this variable, it is used only once

File: features/test.feature [hunk 20]
```diff
@@ -0,0 +1,25 @@
+Feature: testing an Exosphere application
```
Comment (kevgo) [inline features/test.feature]: Why does this file get added here? Do you need to sync the current branch?

File: src/commands/run/service-runner.ls [hunk 21]
```diff
@@ -11,26 +12,52 @@ require! {
 
 class ServiceRunner extends EventEmitter
 
-  (@name, @config) ->
+  (@name, @config, @watcher) ->
     @service-config = yaml.safe-load fs.readFileSync(path.join(@config.root, 'service.yml'), 'utf8')
     @service-config.messages.sends ?= []
     @service-config.messages.receives ?= []
+    @watcher = chokidar.watch @config.root , {
+                              ignoreInitial: true,
```
Comment (kevgo) [inline src/commands/run/service-runner.ls]: Use kebab-case here: `ignore-initial`

File: src/commands/run/service-runner.ls [hunk 22]
```diff
@@ -11,26 +12,52 @@ require! {
...
+                              ignored: /.*\/node_modules\/.*/ }
+      ..on 'add', (watch-path) ~>
+        console.log "exo run: restarting service '#{@name}'"
```
Comment (kevgo) [inline src/commands/run/service-runner.ls]: Maybe even include the watch-path here: "exo run: restarting service '#{@name}' because '#{watch-path} was created'

File: src/commands/run/service-runner.ls [hunk 23]
```diff
@@ -11,26 +12,52 @@ require! {
...
+        ..on 'ended', (exit-code, killed) ~> throw new Error "Service '#{@name}' crashed" unless killed
         ..wait @service-config.startup['online-text'], ~>
           @emit 'online', @name
           done!
 
+  _restart: ~>
+    @server-process
+      ..kill!
+      ..on 'ended', ~>
+        @server-process = new ObservableProcess(@_create-command(@service-config.startup.command)
```
Comment (kevgo) [inline src/commands/run/service-runner.ls]: I assume this code exists somewhere already, to start the service in the first place. Please avoid code duplication here. Refactor this piece into a separate method and call it from both places.

File: templates/add-service/htmlserver-express-es6/package.json [hunk 24]
```diff
@@ -1,5 +1,5 @@
 {
-  "name": "_____serviceName_____",
+  "name": "spacetweet-webserver",
```
Comment (kevgo) [inline templates/add-service/htmlserver-express-es6/package.json]: This change should be reverted. This is a template.

File: templates/add-service/htmlserver-express-livescript/package.json [hunk 25]
```diff
@@ -1,5 +1,5 @@
 {
-  "name": "_____serviceName_____",
+  "name": "spacetweet-webserver",
```
Comment (kevgo) [inline templates/add-service/htmlserver-express-livescript/package.json]: Revert this change

File: templates/add-service/htmlserver-express-es6/package.json (hunk 24 above)
Comment (hugobho) [inline templates/add-service/htmlserver-express-es6/package.json]: Must've been a mistake when merging.

File: features/test.feature (hunk 20 above)
Comment (hugobho) [inline features/test.feature]: I was having some troubles with git. This file should be deleted right? It should exist under features/test/test.feature

**Comment by hugobho** [comment]
@kevgo

File: example-apps/external-service/test/application.yml [hunk 26]
```diff
@@ -0,0 +1,7 @@
+name: Example application
```
Comment (kevgo) [inline example-apps/external-service/test/application.yml]: Let's have the directory `external-service` contain two directories:
- `app`: contains application.yml
- `service`: contains a simple service

File: example-apps/external-service/web-server/package.json [hunk 27]
```diff
@@ -0,0 +1,18 @@
+{
+  "name": "test-web-server",
```
Comment (kevgo) [inline example-apps/external-service/web-server/package.json]: "name": "service"

File: example-apps/external-service/web-server/server.ls [hunk 28]
```diff
@@ -0,0 +1,18 @@
+require! {
+  'http'
+  'exorelay' : ExoRelay
+}
+
+
+exorelay = new ExoRelay exocom-port: 8000, service-name: 'web'
```
Comment (kevgo) [inline example-apps/external-service/web-server/server.ls]: No need to spin up an exorelay here. You don't even use it.

File: example-apps/external-service/web-server/server.ls [hunk 29]
```diff
@@ -0,0 +1,18 @@
...
+  ..on 'online', (port) -> console.log "web service exorelay online at port #{port}"
+  ..on 'error', (err) -> console.log "web service exorelay encountered error: #{err}"
+  ..listen 8002
+
```
Comment (kevgo) [inline example-apps/external-service/web-server/server.ls]: This whole service could just print "starting service". That's all we need here.

File: features/steps/steps.ls [hunk 30]
```diff
@@ -42,6 +42,26 @@ module.exports = ->
...
+    @checkout-app @app-name.split(path.sep)[0]
+    app-dir := path.join process.cwd!, 'tmp', @app-name
+    @setup-app @app-name, ~>
+      @process = new ObservableProcess(path.join(process.cwd!, 'bin', 'exo run'),
+                                      cwd: app-dir,
+                                      stdout: dim-console.process.stdout
+                                      stderr: dim-console.process.stderr)
+        ..wait 'application ready', done
+
+  @Given /^a running "([^"]*)" application inside the "([^"]*)" directory$/, timeout: 600_000, (@app-name, dir-name, done) ->
```
Comment (kevgo) [inline features/steps/steps.ls]: This step seems unused and can be deletede

File: features/steps/steps.ls [hunk 31]
```diff
@@ -201,6 +221,19 @@ module.exports = ->
...
 
 
+  @When /^a file "([^"]*)" gets created$/ (new-filename) ->
+    new-file-path = path.join process.cwd!, 'tmp', @app-name, new-filename
+    fs.open-sync new-file-path, 'w'
+
+  @When /^the file "([^"]*)" gets changed$/ (filename) ->
+    file-path = path.join process.cwd!, 'tmp', @app-name, filename
+    file-data = "var hello = \"hello\""
+    fs.append-file-sync file-path, file-data
```
Comment (kevgo) [inline features/steps/steps.ls]: Temporary variables are a code smell. You can inline it like this:

```ls
fs.append-file-sync path.join(process.cwd!, 'tmp', @app-name, filename),
                    'content'

File: src/commands/run/service-runner.ls [hunk 32]
```diff
@@ -11,26 +12,48 @@ require! {
 
 class ServiceRunner extends EventEmitter
 
-  (@name, @config) ->
+  (@name, @config, @watcher) ->
     @service-config = yaml.safe-load fs.readFileSync(path.join(@config.root, 'service.yml'), 'utf8')
     @service-config.messages.sends ?= []
     @service-config.messages.receives ?= []
+    @watcher = chokidar.watch @config.root , {
+                              ignore-initial: true,
```
Comment (kevgo) [inline src/commands/run/service-runner.ls]: `yes` instead of `true`

File: src/commands/run/service-runner.ls [hunk 33]
```diff
@@ -11,26 +12,48 @@ require! {
...
+  _start-process: (done) ~>
+    @server-process = new ObservableProcess(@_create-command(@service-config.startup.command)
+                          cwd: @config.root,
+                          env: @config
+                          stdout: {@write} 
+                          stderr: {@write})
+      ..on 'ended', (exit-code, killed) ~> throw new Error "Service '#{@name}' crashed" unless killed
+      ..wait @service-config.startup['online-text'], ~>
+        @emit 'online', @name
+        console.log \here
```
Comment (kevgo) [inline src/commands/run/service-runner.ls]: delete this lin

File: templates/add-service/htmlserver-express-es6/app/index.js [hunk 34]
```diff
@@ -1,6 +1,7 @@
 // This is the main server file.
 //
 // It parses the command line and instantiates the two servers for this app:
+const async = require('async')
```
Comment (kevgo) [inline templates/add-service/htmlserver-express-es6/app/index.js]: Please extract this change and the one in the next file into a separate PR

File: features/steps/steps.ls (hunk 30 above)
Comment (hugobho) [inline features/steps/steps.ls]: This is used in the auto_restart.feature file

File: templates/add-service/htmlserver-express-es6/app/index.js (hunk 34 above)
Comment (hugobho) [inline templates/add-service/htmlserver-express-es6/app/index.js]: I think I removed this line in a separate PR (update async), but it got re-added when I had some trouble merging this branch with the master

**Comment by hugobho** [comment]
@kevgo

File: features/auto_restart.feature [hunk 35]
```diff
@@ -0,0 +1,27 @@
...
+  I want that any changes I make to the source code get loaded immediately into the runtime
+  So that I don't waste time manually restarting services all the time, or setting up service-specific auto-reloading.
+
+  Rules:
+  - anytime a file in a service directory is created, updated, or deleted, "exo run" restarts the respective service
+  - when a service is restarted, the logger outputs: "exo run: restarting service '<service name>'"
+
+  Scenario: file change in an internal service
+    Given a running "test" application
+    When a file "web-server/new-file.js" gets created
```
Comment (kevgo) [inline features/auto_restart.feature]: Ideally each scenario checks only one condition. I think it's also more precise if we break up each test case into its own scenario:

```cucumber
Scenario: creating a file in an internal service
Scenario: updating a file in an internal service
Scenario: deleting a file in an internal service
Scenario: creating a file in an external service
...

File: features/steps/steps.ls [hunk 36]
```diff
@@ -307,6 +329,11 @@ module.exports = ->
     done!
 
 
+  @Then /^the "([^"]*)" service restarts( again)?$/, (service, again, done) ->
```
Comment (kevgo) [inline features/steps/steps.ls]: You don't use `again`, so tell the regex to ignore it:

```ls
@Then /^the "([^"]*)" service restarts(?: again)?$/, (service, done) ->

File: features/steps/steps.ls [hunk 37]
```diff
@@ -307,6 +329,11 @@ module.exports = ->
...
+    @process.reset-output-streams!
+    @process.wait "exorun  '#{service}' is running using exorelay port", done
```
Comment (kevgo) [inline features/steps/steps.ls]: exorun should say something like "restarting #{service} because #{file} ..."

File: src/commands/run/service-runner.ls [hunk 38]
```diff
@@ -15,22 +16,43 @@ class ServiceRunner extends EventEmitter
...
-                            stderr: {@write})
-        ..on 'ended', ~> throw new Error "Service '#{@name}' crashed"
-        ..wait @service-config.startup['online-text'], ~>
-          @emit 'online', @name
-          done!
+      @_start-process done
 
+  _restart: ~>
+    @server-process
+      ..kill!
```
Comment (kevgo) [inline src/commands/run/service-runner.ls]: register the callback first, then call kill. Event handlers are called synchronously.

File: src/commands/run/service-runner.ls [hunk 39]
```diff
@@ -15,22 +16,43 @@ class ServiceRunner extends EventEmitter
...
+      ..on 'ended', ~>
```
Comment (kevgo) [inline src/commands/run/service-runner.ls]: ..on 'ended', @_start-process

File: features/auto_restart.feature (hunk 35 above)
Comment (hugobho) [inline features/auto_restart.feature]: @kevgo 

Is there a way in Cucumber to run one step once and then multiple scenarios against it? For instance, set up a running test application once, then just run each of these scenarios that create, update, delete a file. It's taking a long time to have to boot up a running "test" application. I'm also running into ports being in use, I assume from repeatedly starting the same application

Comment (kevgo) [inline features/auto_restart.feature]: I'd use a [scenario outline](https://github.com/cucumber/cucumber/wiki/Scenario-Outlines) here.

**Comment by trushton** [comment]
Closing this PR as major changes have been made since its last update. Work on this issue will be continued under the [tr-restart-services](https://github.com/Originate/exo-run/tree/tr-restart-services) branch and issue [Originate/exosphere-sdk#165](https://github.com/Originate/exosphere-sdk/issues/165).
